presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpiller.java

```diff
@@ -168,8 +168,9 @@ private void writePages(Iterator<Page> pageIterator)
 
         try {
             InputStream input = closer.register(targetFile.newInputStream());
-            Iterator<Page> pages = PagesSerdeUtil.readPages(serde, new InputStreamSliceInput(input, BUFFER_SIZE));
-            return closeWhenExhausted(pages, input);
+            Iterator<Page> deserializedPages = PagesSerdeUtil.readPages(serde, new InputStreamSliceInput(input, BUFFER_SIZE));
+            Iterator<Page> compactPages = new SpillerUtil.CompactPageIterator(deserializedPages);
```

Comment:
nit:

```
 Iterator<Page> compactedPages = transform(deserializedPages, page -> {
                page.compact();
                return page;
            });
```

or simply `Iterator<Page> compactedPages = transform(deserializedPages, Page::compact)` if you change the `Page::compact` implementation to return `this` as in other methods in the `Page` interface.